src/inmanta/agent/forking_executor.py

```diff
@@ -153,54 +205,111 @@ def _force_stop(self, grace_time: float) -> None:
+        self.owner._child_closed(self)
 
-class MPManager:
-    def __init__(self, log_folder: str, inmanta_log_level: str = "DEBUG", cli_log: bool = False) -> None:
+    async def join(self, timeout: float) -> None:
+        if self.closed:
+            return
+        try:
+            await asyncio.get_running_loop().run_in_executor(None, functools.partial(self.process.join, timeout))
+            self.process.close()
+            self._set_closed()
+        except ValueError as e:
+            if "process object is closed" in str(e):
+                # process already closed
+                # raises a value error, so we also check the message
+                pass
+            else:
+                raise
+
+
+class MPManager(executor.ExecutorManager[MPExecutor]):
+    def __init__(
+        self,
+        thread_pool: concurrent.futures.thread.ThreadPoolExecutor,
+        environment_manager: executor.VirtualEnvironmentManager,
+        session_gid: uuid.UUID,
+        log_folder: str,
+        storage_folder: str,
+        inmanta_log_level: str = "DEBUG",
+        cli_log: bool = False,
+    ) -> None:
         """
-
+        :param thread_pool:  threadpool to perform work on
+        :param environment_manager: The VirtualEnvironmentManager responsible for managing the virtual environments
+        :param session_gid: agent session id, used to connect to the server, the agent should keep this alive
         :param log_folder: folder to place log files for the executors
+        :param storage_folder: folder to place code files
         :param inmanta_log_level: log level for the executors
         :param cli_log: do we also want to echo the log to std_err
+
         """
+        super().__init__(thread_pool, environment_manager)
         self.children: list[MPExecutor] = []
         self.log_folder = log_folder
+        self.storage_folder = storage_folder
         os.makedirs(self.log_folder, exist_ok=True)
+        os.makedirs(self.storage_folder, exist_ok=True)
         self.inmanta_log_level = inmanta_log_level
         self.cli_log = cli_log
+        self.session_gid = session_gid
 
     @classmethod
     def init_once(cls) -> None:
         try:
             multiprocessing.set_start_method("forkserver")
-            multiprocessing.set_forkserver_preload(["inmanta", "inmanta.config"])
+            # Load common modules
+            # Including this one
+            multiprocessing.set_forkserver_preload(["inmanta.config", __name__])
         except RuntimeError:
             # already set
             pass
 
+    async def create_executor(self, venv: executor.ExecutorVirtualEnvironment, executor_id: executor.ExecutorId) -> MPExecutor:
+        # entry point from parent class
+        executor = await self.make_child_and_connect(executor_id.agent_name)
+        storage_for_blueprint = os.path.join(self.storage_folder, "code", executor_id.blueprint.generate_blueprint_hash())
+        os.makedirs(storage_for_blueprint)
```

Comment:
no, it serves/may serve as a cache as well I think

Reply:
but it didn't handle existing dirs correctly, thx

Reply:
But then we still have the problem where inter-module dependencies may cause stale code to be present in the code directory. I though that solving that problem was one of the reasons why we are making these changes. But indeed, it would require starting from a clean cache on every update.

Reply:
(But I think we can postpone this discussion. Maybe we have to create a ticket for this and discuss it on the planning meeting)

Reply:
it can't happen, because the name of the folder is the hash of the files, ...